show token scopes on auth status

mntlty opened the pull request:
Closes https://github.com/cli/cli/issues/6520

The current query to get the logged in user uses the graphql package https://pkg.go.dev/github.com/cli/shurcooL-graphql which does not expose the http headers in the response, which is where the scopes information is found. I also looked at https://github.com/cli/cli/blob/37e343a80524ba437fc18d66876f1739f97f785f/api/client.go#L14 but ran into the same issue, the response struct is not returned to the caller.

I also added some documentation to the `httpmock.REST` command.

Output looks like:

![Screen Shot 2022-10-31 at 4 40 49 PM](https://user-images.githubusercontent.com/1149246/199130513-abc96787-dd3c-42db-abe1-e11e9603066a.png)

In the case that no scopes are found, the output will say
```
X Token Scopes: None found
```

There are different prefixes for token types https://github.blog/2022-10-18-introducing-fine-grained-personal-access-tokens-for-github/

Server-to-server and Fine grained permissions tokens do not contain the `"X-Oauth-Scopes"` header, which would make the output above confusing. In those cases, we show a warning to the user.

For a fine-grained token (prefixed with `github_pat_`):
```
github.com
  X github.com: api call failed: non-200 OK status code: 401 Unauthorized body: "{\"message\":\"Personal access tokens with fine grained access do not support the GraphQL API\",\"documentation_url\":\"https://docs.github.com/graphql/guides/forming-calls-with-graphql#authenticating-with-graphql\"}"
  ✓ Logged in to github.com as  (GH_TOKEN)
  ✓ Git operations for github.com configured to use ssh protocol.
  ✓ Token: *******************
  ! Cannot determine scopes for fine-grained token.
```

For a server-to-server token (prefixed with `ghs_`):  
```
github.com
  ✓ Logged in to github.com as mntlty-test[bot] (GH_TOKEN)
  ✓ Git operations for github.com configured to use ssh protocol.
  ✓ Token: *******************
  ! Cannot determine scopes for server-to-server token.
```

mntlty committed:
show token scopes on auth status

mntlty commented on pkg/cmd/auth/status/status_test.go:
```diff
@@ -143,13 +143,14 @@ func Test_statusRun(t *testing.T) {
+
 			},
-			wantErrOut: regexp.MustCompile(`(?s)Logged in to github.com as.*tess.*Logged in to joel.miller as.*tess`),
+			wantErrOut: regexp.MustCompile(`(?s)Logged in to github.com as.*tess.*Token Scopes: read:org, read:packages.*Logged in to joel.miller as.*tess.*X Token Scopes: None found`),
```
tests output for scopes found and not found

mislav commented on pkg/cmd/auth/status/status.go:
```diff
@@ -164,3 +171,26 @@ func statusRun(opts *StatusOptions) error {
 }
+
+func userAndScopes(httpClient *http.Client) (string, string, error) {
+	url := fmt.Sprintf("%suser", ghinstance.RESTPrefix(ghinstance.Default()))
```
Hi, the `user` endpoint isn't available when using the generated GITHUB_TOKEN in Actions. The API returns an error: `resource not available by integration` and the command aborts.

On the other hand, fetching the viewer via GraphQL works and returns `Logged in to github.com as github-actions[bot]`.

Also note that we already have a mechanism of getting scopes: https://github.com/cli/cli/blob/fa2223129b536efad8704a66c270630b47dc8b1d/pkg/cmd/auth/shared/oauth_scopes.go#L34

mntlty replied:
@mislav thank you! I wasn't thrilled with the implementation I had, this one is much better.

In the earlier line in the command https://github.com/cli/cli/blob/6bf194f63e3cf6de0ea69d99fb0d3c9a726f7c50/pkg/cmd/auth/status/status.go#L100 we are fetching the scopes and discarding them, which necessitates a second API call to get them. What do you think about changing the signature of https://github.com/cli/cli/blob/6bf194f63e3cf6de0ea69d99fb0d3c9a726f7c50/pkg/cmd/auth/shared/oauth_scopes.go#L63 to return `(string, error)` where string is the list of scopes from the api?

It would save an extra api call and make the tests less repetitive. A quick code search shows this function is only called in a few places.

mntlty replied:
One thing I noticed in testing, in https://github.com/cli/cli/blob/6bf194f63e3cf6de0ea69d99fb0d3c9a726f7c50/pkg/cmd/auth/status/status.go#L124 and the one I added below it https://github.com/cli/cli/blob/6bf194f63e3cf6de0ea69d99fb0d3c9a726f7c50/pkg/cmd/auth/status/status.go#L129 we don't bubble up errors to the command, we add them to the output message. 

This meant that in some of the tests I would see unmatched http stubs in the output but the test conditions of not having an error and matching parts of the regular expression message we ok, and so the test passed.

samcoe replied:
@mntlty I like the idea of keeping it to just a single api request for the scope headers. What do you think about creating a new function `HeaderHasMinimumScopes(scopeHeader string) error` that would be called from `HasMinimumScopes` and also the `auth status` command with the extracted scopes header so that we don't have to change the function signature of `HasMinimumScopes`?

As for the tests, we probably should add `defer reg.Verify(t)` to these tests so that they fail on unmatched http stubs.

mntlty replied:
@samcoe thanks for the feedback! I think this version is nicer 😄 - suggest reviewing with whitespace off due to an `else` clause I added to keep the existing structure. I left some comments in the PR, I think `func statusRun` could really use a refactor, which unfortunately I can't commit to at this time. Hopefully that won't hold back this change, which I think is a really helpful one.

I was surprised to see that `reg.Verify(t)` was already there, and without the `defer` it wasn't having the intended effect of failing unmatched stubs and other silent errors. I wonder if making that less error-prone in some way is something for the team to consider as a future improvement. 🙏

mntlty committed:
use shared.GetScopes

mntlty committed:
only make a single api call to get scopes

mntlty committed:
add defer to reg.Verify(t) and update tests

mntlty commented on pkg/cmd/auth/shared/oauth_scopes_test.go:
```diff
@@ -11,6 +11,54 @@ import (
 )
 
 func Test_HasMinimumScopes(t *testing.T) {
+	tests := []struct {
```
broke up the existing tests from `Test_HasMinimumScopes` to a single pass and fail with `httpmock`, and copied the rest below as unit tests to `Test_HeaderHasMinimumScopes`

mntlty committed:
do not return error mark it as failed

mntlty commented on pkg/cmd/auth/status/status.go:
```diff
@@ -97,48 +97,61 @@ func statusRun(opts *StatusOptions) error {
+		if err != nil {
+			failed = true
+			addMsg("%s %s: api call failed: %s", cs.Red("X"), hostname, err)
+		} else {
```
adding this else pushed everything to the right, but I think is necessary to keep with the current pattern of setting `failed = true` and continuing execution. This function would benefit from a refactor, unfortunately I'm not able to take that on at this time

mntlty commented on pkg/cmd/auth/status/status.go:
```diff
@@ -97,48 +97,61 @@ func statusRun(opts *StatusOptions) error {
+				apiClient := api.NewClientFromHTTP(httpClient)
+				username, err := api.CurrentLoginName(apiClient, hostname)
+				if err != nil {
+					addMsg("%s %s: api call failed: %s", cs.Red("X"), hostname, err)
```
I believe this should have `failed = true` but don't see a test covering it either way. deferring to the maintainers to make a call here

samcoe committed:
Small polish/style changes

samcoe reviewed (approved):
LGTM! Thanks for the contribution. I pushed a small 💅 commit to cleanup some small style nits I had.

mislav commented on pkg/cmd/auth/status/status.go:
```diff
@@ -134,6 +144,12 @@ func statusRun(opts *StatusOptions) error {
+			if scopesHeader != "" {
+				addMsg("%s Token Scopes: %s", cs.SuccessIcon(), scopesHeader)
+			} else {
+				addMsg("%s Token Scopes: None found", cs.Red("X"))
```
Rendering this as an error in red color might be misleading for Integration tokens and PAT v2, which have a concept of "permissions" but not scopes, and thus do not render any `x-oauth-scopes` response header.

Maybe it's best to just not render information about "scopes" in that case and instead print the type of token used?

mntlty replied:
@mislav thanks, I made the changes and updated the PR body to reflect if we don't know the token type and offer more information. If you think it's clearer, I can remove it entirely in those cases. 

Thinking about this, I wonder if 
- the current output is perhaps not correct with a server-to-server-token, as we can't know if the token has the scopes needed for cli operations
- we should modify the message when we don't know the user's login (see the output from a fine-grained token)
- we should modify some of the other output when using a fine-grained token, where we also show an error message

I feel like these may be out of scope from the original issue, but wanted to bring them up.

mislav reviewed (changes requested):
Almost there! Thank you @mntlty

mntlty committed:
show scopes warning message for fine-grained and server-to-server tokens

mislav committed:
auth status tweaks

- show path to config file
- display token prefix
- refine how missing scopes are shown

mislav committed:
Merge remote-tracking branch 'origin' into mntlty/user_scopes

mislav reviewed (approved):
Thanks @mntlty for the contribution and @samcoe for reviewing.

I've pushed some changes:
- When a token for a specific host is invalid (i.e. results in a HTTP 4xx), this restores the original error message that prints: `The {host} token in {source} is no longer valid.`
- Unless the token comes from environment variables, the value for `{source}` is now a file path to `hosts.yml` instead of a hardcoded string `oauth_token` (which might be a bug in go-gh)
- When printing a masked token, print the token prefix like `ghp_` unmasked so that the user can inspect which type of token they are using
- Only print scopes information for “classic” PATs and OAuth tokens
- `gh auth status` tests now assert exact command output instead of matching with regexp

mntlty commented:
@mislav can this be merged?

mislav commented:
Yes. I guess that because the scope of my added changes, I wanted to give you a chance to weigh in.
